ICAFIX/ApplyAutoReclean.sh

```diff
@@ -0,0 +1,236 @@
+matlab_argarray=("$StudyFolder" "$Subject" "$fMRIListName" "$subjectExpectedTimepoints" "$HighPass" "$fMRIResolution" "$CorticalParcellationFile" "$SubCorticalParcellationFile" "$WMLabelFile" "$CSFLabelFile")
+
+case "$MatlabMode" in
+    (0)
+        matlab_cmd=("$this_script_dir/Compiled_computeRecleanFeatures/run_computeRecleanFeatures.sh" "$MATLAB_COMPILER_RUNTIME" "${matlab_argarray[@]}")
+        log_Msg "running compiled matlab command: ${matlab_cmd[*]}"
+        "${matlab_cmd[@]}"
+        ;;
+    (1 | 2)
+        #reformat argument array so matlab sees them as strings
+        matlab_args=""
+        for thisarg in "${matlab_argarray[@]}"
+        do
+            if [[ "$matlab_args" != "" ]]
+            then
+                matlab_args+=", "
+            fi
+            matlab_args+="'$thisarg'"
+        done
+        
+        matlabcode="
+            addpath('$HCPPIPEDIR/global/matlab');
+            addpath('$HCPCIFTIRWDIR');
+            addpath('$this_script_dir/scripts');
+            addpath('$this_script_dir');
+            computeRecleanFeatures($matlab_args);"
+
+        log_Msg "running matlab code: $matlabcode"
+        "${matlab_interpreter[@]}" <<<"$matlabcode"
+        echo
+        ;;
+esac
+
+# hardcoded
+FixProbThresh="10"
+
+# inference each subjects under python environment
+for fMRIName in ${fMRINameToUse} ; do
+    RecleanFeaturePath="${StudyFolder}/${Subject}/MNINonLinear/Results/${fMRIName}/${fMRIName}_hp${HighPass}.ica/fix_reclean_features.csv"
+    FixProbPath="${StudyFolder}/${Subject}/MNINonLinear/Results/${fMRIName}/${fMRIName}_hp${HighPass}.ica/fix_prob.csv"
+    PredictionResult="${StudyFolder}/${Subject}/MNINonLinear/Results/${fMRIName}/${fMRIName}_hp${HighPass}.ica"
+
+    # ReclassifyAsSignalTxt="${PredictionResult}/ReclassifyAsSignalRecleanVote${VoteThresh}.txt"
+    # ReclassifyAsNoiseTxt="${PredictionResult}/ReclassifyAsNoiseRecleanVote${VoteThresh}.txt"
+
+    # ReclassifyAsSignalTxt="${PredictionResult}/${ReclassifyAsSignalFile}"
+    # ReclassifyAsNoiseTxt="${PredictionResult}/${ReclassifyAsNoiseFile}"
+
+    ReclassifyAsSignalTxt="${StudyFolder}/${Subject}/MNINonLinear/Results/${fMRIName}/${ReclassifyAsSignalFile}"
+    ReclassifyAsNoiseTxt="${StudyFolder}/${Subject}/MNINonLinear/Results/${fMRIName}/${ReclassifyAsNoiseFile}"
+
+    pythonCode=(
+        "singularity exec --bind /media $PythonSingularity python3 $HCPPIPEDIR/ICAFIX/scripts/RecleanClassifierInference.py"
+        "--input_csv=$RecleanFeaturePath"
+        "--input_fix_prob_csv=$FixProbPath"
+        "--fix_prob_threshold=$FixProbThresh"
+        "--trained_folder=$ModelFolder"
+        "--model=$ModelToUse"
+        "--output_folder=$PredictionResult"
+        "--voting_threshold=$VoteThresh"
+        "--reclassify_as_signal_file=$ReclassifyAsSignalTxt"
+        "--reclassify_as_noise_file=$ReclassifyAsNoiseTxt"
+    )
+
+    cmd="${pythonCode[*]}"
+    log_Msg "Run python inference: $cmd"
+    eval "$cmd"
```

Comment:
Please don't use `eval` unless there is no other choice, it will do bad things with any special characters that might be in a filename.  `"${pythonCode[@]}"` is the right way to run this, but you need to have the spaces between `singularity exec --bind`, etc unquoted inside the `()`, so that bash knows they should be separate elements in the array.

Reply:
Also, binding "/media" is somewhat specific to our setup.  This kind of thing is one reason I think the pipeline scripts shouldn't be trying to run a container internally, instead the container setup and run should be outside the pipeline, and the pipeline shouldn't responsible for bind mounts, etc.

Reply:
Or, we can add an argument to represent the "/media" if user wants to customize it like what we did. 

Putting the pipeline inside a container sounds like the approach what Qunex is trying to do. Or, shall we have a conda environment designed for all the python scripts in the HCP pipeline?

Reply:
That is my point, qunex is doing it, the pipelines aren't calling a container.  For a conda environment, I think that is also something to be done before calling the pipeline, not something for the pipelines themselves to mess with.

Put simply, I think the pipeline itself should just have a bare call to python, expecting it to work in the current environment, whatever that may be.  conda/container/vm/whatever is something that can be used to help set things up for the pipeline to work, I don't think it should be the other way around.  Matlab is an unfortunate exception where the licensing/cost prevents code we wrote in matlab from being feasibly runnable in many circumstances, so the "compiled matlab" stuff is a hack to deal with our choice to use matlab.  No other language we use should have special code to run things through a different mechanism.